docs/enhancements/auto-return-agent-to-infra-env.md

```diff
@@ -0,0 +1,140 @@
+in installed clusters to run an agent. When appropriate this agent will
+reconfigure the host to boot into a discovery image, allowing the host to
+be made available for installation without the user manually interacting with
+a management system to reboot the host into the discovery image.
+
+## Motivation
+
+A user creating and managing clusters using the assisted-service may not have
+access to the systems used to configure and boot hosts. This is especially
+true in bare-metal use cases where a BareMetalHost is not present. The
+assisted-service can make this use case easier by handling the reprovisioning
+process.
+
+### Goals
+
+* Reconfigure a host to boot the discovery image as a part of unbinding from a
+  cluster without a BareMetalHost
+* Allow for a node to be gracefully removed from an installed cluster
+
+### Non-Goals
+
+* Automate graceful removal of a node from an existing cluster
+* Booting a node that is unresponsive or not currently part of an installed
+  cluster
+
+## Proposal
+
+When a host should be returned to the InfraEnv use the kubeconfig created when
+the cluster was installed to run an agent on the host. This agent will handle
+downloading the required artifacts and rebooting the host into the discovery
+process.
+
+### Timeline
+
+1. User triggers an action which will prepare a host to be unbound
+  * Assisted service will create a MachineConfig to run our agent on the host as
+    a systemd service.
+  * The agent will download the InfraEnv's kernel and initrd (including
+    discovery ignition) to the host's `/boot` partition.
+  * The agent will create a GRUB boot entry to boot into the discovery kernel
+    and initrd.
+2. User (or other automation) removes the node from the cluster
+  * Drain existing workloads and delete the node
+3. User unbinds the host from the cluster using the assisted-service API
+  * This triggers the agent to reboot the host into the discovery image
+4. Host is registered as `known-unbound` in the InfraEnv and is ready to be used
+   in new clusters
+
+### User Stories
+
+#### Story 1
+
+As a cluster creator I want to use the assisted service to repurpose hosts 
+from existing clusters without requiring access to the host management 
+infrastructure (BMC, hypervisor, etc).
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. A way to query for URLs to the iPXE boot artifacts over the REST API should be
+implemented. Currently these URLs are only exposed through the iPXE script and
+in the InfraEnv CR, but the agent running on a node will need access to these
+artifacts.
+  * This can be added to the existing payload returned when querying an InfraEnv
+  * Alternatively this would need a new API endpoint
+    * `/v2/infra-envs/{infra_env_id}/boot-artifacts`?
+2. The agent will likely require two changes/additions
+  * A step to query the boot artifacts and write the boot entry
+  * A step to reboot the host into the new boot entry
```

Comment:
The agent won't know. The service will tell it when it should do the reboot.
In turn the service knows because the user said it was safe to unbind the host.

Reply:
something is missing here for me as well, currently unbind mean that user removed the cluster reference, i'm not sure i understand the flow here.

Reply:
Talked to @filanov about this.
I'm working on adding some details about the changes to the host state machine that I have in mind.